src/react/Reconciler.js

```diff
@@ -0,0 +1,332 @@
+    // as we don't support class components yet, throw a fatal error
+    throw new ExpectedBailOut("class components not yet supported");
+  } else {
+    // otherwise it's a functional component, where the first paramater of the function is "props" (if it exists)
+    if (componentType.$FormalParameters.length > 0) {
+      let firstParam = componentType.$FormalParameters[0];
+      if (t.isIdentifier(firstParam)) {
+        propsName = ((firstParam: any): BabelNodeIdentifier).name;
+      }
+      let propsTypeAnnotation = firstParam.typeAnnotation !== undefined && firstParam.typeAnnotation;
+      // we expect that if there's a props paramater, it should always have Flow annotations
+      if (!propsTypeAnnotation) {
+        throw new ExpectedBailOut(`root component missing Flow type annotations for the "props" paramater`);
+      }
+      propTypes = flowAnnotationToObjectTypeTemplate(propsTypeAnnotation);
+    }
+  }
+  return createAbstractObject(realm, propsName, propTypes);
+}
+
+function getInitialContext(realm: Realm, componentType: ECMAScriptSourceFunctionValue): ObjectValue | AbstractValue {
+  let contextName = null;
+  let contextTypes = null;
+  if (valueIsClassComponent(realm, componentType)) {
+    // it's a class component, so we need to check the type on for context of the component prototype
+    // as we don't support class components yet, throw a fatal error
+    throw new ExpectedBailOut("class components not yet supported");
+  } else {
+    // otherwise it's a functional component, where the second paramater of the function is "context" (if it exists)
+    if (componentType.$FormalParameters.length > 1) {
+      let secondParam = componentType.$FormalParameters[1];
+      if (t.isIdentifier(secondParam)) {
+        contextName = ((secondParam: any): BabelNodeIdentifier).name;
+      }
+      let contextTypeAnnotation = secondParam.typeAnnotation !== undefined && secondParam.typeAnnotation;
+      // we expect that if there's a context param, it should always have Flow annotations
+      if (!contextTypeAnnotation) {
+        throw new ExpectedBailOut(`root component missing Flow type annotations for the "context" paramater`);
+      }
+      contextTypes = flowAnnotationToObjectTypeTemplate(contextTypeAnnotation);
+    }
+  }
+  return createAbstractObject(realm, contextName, contextTypes);
+}
+
+class Reconciler {
+  constructor(
+    realm: Realm,
+    moduleTracer: ModuleTracer,
+    statistics: ReactStatistics,
+    reactSerializerState: ReactSerializerState
+  ) {
+    this.realm = realm;
+    this.moduleTracer = moduleTracer;
+    this.statistics = statistics;
+    this.reactSerializerState = reactSerializerState;
+  }
+
+  realm: Realm;
+  moduleTracer: ModuleTracer;
+  statistics: ReactStatistics;
+  reactSerializerState: ReactSerializerState;
+
+  render(componentType: ECMAScriptSourceFunctionValue): Effects {
+    return this.realm.wrapInGlobalEnv(() =>
+      this.realm.evaluateForEffects(() => {
```

Comment:
You could use the return value of this to detect if there are any mutations on objects other than newly created ones. Then log those to the error logger. That'll help us track violations in components. :)

Reply:
I'll add a TODO